tests/UnifiedSpecTests/Context.php

```diff
@@ -0,0 +1,395 @@
+
+            $id = $def->id ?? null;
+            assertInternalType('string', $id);
+            assertArrayNotHasKey($id, $this->entityMap);
+
+            switch ($type) {
+                case 'client':
+                    $this->entityMap[$id] = $this->createClient($def);
+                    break;
+
+                case 'database':
+                    $this->entityMap[$id] = $this->createDatabase($def);
+                    break;
+
+                case 'collection':
+                    $this->entityMap[$id] = $this->createCollection($def);
+                    break;
+
+                default:
+                    throw new LogicException('Unsupported entity type: ' . $type);
+            }
+        }
+    }
+
+    public function getEntityMap() : EntityMap
+    {
+        return $this->entityMap;
+    }
+
+    public function getInternalClient() : Client
+    {
+        return $this->internalClient;
+    }
+
+    public function prepareOperationArguments(array $args) : array
+    {
+        if (array_key_exists('readConcern', $args)) {
+            assertInternalType('object', $args['readConcern']);
+            $args['readConcern'] = self::prepareReadConcern($args['readConcern']);
+        }
+
+        if (array_key_exists('readPreference', $args)) {
+            assertInternalType('object', $args['readPreference']);
+            $args['readPreference'] = self::prepareReadPreference($args['readPreference']);
+        }
+
+        if (array_key_exists('session', $args)) {
+            assertInternalType('string', $args['session']);
+            assertArrayHasKey($args['session'], $this->entityMap);
+            $session = $this->entityMap[$args['session']];
+            assertInstanceOf(Session::class, $session);
+            $args['session'] = $session;
+        }
+
+        if (array_key_exists('writeConcern', $args)) {
+            assertInternalType('object', $args['writeConcern']);
+            $args['writeConcern'] = self::prepareWriteConcern($args['writeConcern']);
+        }
+
+        return $args;
+    }
+
+    public function startEventObservers()
+    {
+        foreach ($this->eventObserversByClient as $eventObserver) {
+            $eventObserver->start();
+        }
+    }
+
+    public function stopEventObservers()
+    {
+        foreach ($this->eventObserversByClient as $eventObserver) {
+            $eventObserver->stop();
+        }
+    }
+
+    private static function assertHasOnlyKeys($arrayOrObject, array $keys)
+    {
+        assertThat($arrayOrObject, logicalOr(isType('array'), isType('object')));
+        $diff = array_diff_key((array) $arrayOrObject, array_fill_keys($keys, 1));
+        assertEmpty($diff, 'Unsupported keys: ' . implode(',', array_keys($diff)));
+    }
+
+    private function createClient(stdClass $o) : Client
+    {
+        self::assertHasOnlyKeys($o, ['id', 'uriOptions', 'useMultipleMongoses', 'observeEvents', 'ignoreCommandMonitoringEvents']);
```

Comment:
Nice way to be defensive about what keys we support. Would it make sense to expand the error message a bit to give pointers about where the unsupported keys occur? So maybe "Unsupported keys in client entity: ..." as opposed to the current "Unsupported keys: ...".

Reply:
The stack trace of the assertion failure should tell us the exact context, so I don't think it's worth the trouble. In the case of `prepareCollectionOrDatabaseOptions` we wouldn't know the exact database or collection entity anyway.

Generally we should never hit these errors unless we're executing test files that don't comply with the JSON schema, since I used `additionalProperties: false` throughout.

Reply:
That makes sense. In that case, I'm happy to leave this as is.